### ISSUE TITLE
Hook for awaiting moderation note in comment

### ISSUE THREAD

**Issue by iljester** [ClassicPress/ClassicPress#1469]
### Context

I tried to check, but in the Classicpress core there is no filter or action that can customize the moderation waiting message.

I looked in the class class-walker-comment.php and it doesn't exist (line 338).

It would be appropriate to pass this note through an action and then, output throght a filter. See example below.

In class-walker-comment.php, line 341:

```php
/**
 * hook that prints the moderation note
 * Default action `display_comment_moderation_note` in wp-includes/comment-template.php includes a moderation note.
 * 
 * @since CP-2.x.x
 */
do_action('comment_moderation_note', $moderation_note, $comment );
```

In comment-template.php, line 1251:

```php
/**
 * Outputs comment moderation note
 *
 * @param string $moderation_note Moderation note text.
 * @param object $comment Current comment information.
 */
function display_comment_moderation_note( $moderation_note, $comment ) {
	
        $html = '';
	if ( '0' == $comment->comment_approved ) {			
		$html = sprintf( '<em class="comment-awaiting-moderation">%s</em>', $moderation_note );
	}

        /**
         * Filter output for comment moderation note
         * @param $html HTML note string
         * @param $moderation_note note string
         * @param $comment Comment information
         * Since CP-2.x.x
         */
        echo apply_filters( 'get_comment_moderation_note', $html, $moderation_note, $comment );
	
}
add_action('comment_moderation_note', 'display_comment_moderation_note', 10, 2 );
```

Furthermore, by moving the message within the comment_content div container, and not within the footer tags of the commenter's info and the comment date, which I believe creates design problems in the themes.

However, there is a way to get around the lack of action, using a filter, "comment_text" (or get_comment_text), which adds the moderation note inside the comment. But then you need to use javascript or css to hide or remove the duplicate by default.

Reference: class-walker-comment.php, method "html5_comment()" - comment-template.php

Thanks for the attention.

### Possible implemantion

see above or screenshot.

### Possible Solution

see above or screenshot.

### Will you be able to help with the implementation?

For convenience, I also attach the images with modifications:

On wp-includes/class-walker-comment.php (inside html5_comment method)
<img width="1049" alt="Screenshot 2024-06-14 alle 07 46 49" src="https://github.com/ClassicPress/ClassicPress/assets/56711459/efe93e3d-46f2-4069-9d0d-082b1c2c26d2">

On wp-includes/comment-template.php
<img width="847" alt="Screenshot 2024-06-15 alle 05 45 55" src="https://github.com/ClassicPress/ClassicPress/assets/56711459/3942db6a-edb2-4ee7-9826-e009c7d22957">
with the changes.

**Comment by KTS915** [comment]
The comment moderation message is localizable. It just needs translating either via a CP translation (if there is one for a specific language) or via the `gettext` filter (or a plugin that makes use of that filter, e.g. https://wordpress.org/plugins/text-replace/.) The filter approach can also be used to modify the message even if it remains in English.

**Comment by iljester** [comment]
Ok. But it's not the same thing to translate the message or allow different customizations, even with html. The filter serves precisely to allow the developer greater freedom of action. I had thought about translation via gettext, but translation is not really an appropriate area of ​​development. It is only used to localize a theme or plugin into the local language.

**Comment by KTS915** [comment]
If all you want to do is change the text, then using `gettext` is precisely the tool for the job. It even accommodates adding HTML. What "greater freedom of action" are you saying someone might need?

**Comment by ginsterbusch** [comment]
Yeah, translation is just "replace existing message with translated version of message", not "replace whole HTML output with something else". Also a hook is easier to pinpoint, while gettext_(filter) might be reused elsewhere.

**Comment by KTS915** [comment]
If we take that approach, then we'll be adding a filter to every translatable string. So I ask again what else, apart from modifying the text, is a developer going to use such a filter for.

**Comment by ginsterbusch** [comment]
> If we take that approach, then we'll be adding a filter to every translatable string. So I ask again what else, apart from modifying the text, is a developer going to use such a filter for.

Oh, right, didnt realize its filter vs. action hook. Then, yes, an actual action hook would be the more wise "action" (no pun intended).

**Comment by pattonwebz** [comment]
Sounds to me like the need here is more about being able to modify the markup than it is to filter the translated string. It doesn't appear like it's currently possible to modify that markup currently via an action or filter, but it probably is possibly to do by subclassing the walker and overriding the method that produces the output.

Subclassing the walker is less clean of a way than a filter inside of a theme/plugin - but it means no core code changes would be needed.

**Comment by iljester** [comment]
> If we take that approach, then we'll be adding a filter to every translatable string. So I ask again what else, apart from modifying the text, is a developer going to use such a filter for.

It's not a matter of adding a filter to every translatable string. I also completely agree that it can't be done, but in the case of the moderation waiting message, this is a important point in the development of a theme or plugin dedicated to comments, and the developer may deem it appropriate to change not only the text of the message, but the entire output, adding icons, creating javascript events, changing the entire geometry of the message, or simply want to hide it and create one elsewhere in the comment body. I have no idea what a developer would want.

I can tell you my experience: I created a comments plugin in Ajax, and that message bothered me inside the footer tag of the meta author and date of the comment. I've been looking for a hook to remove it. I have not found it. I thought about gettext, but at that point I would have to do a translation process that I didn't intend to do. I also thought about extending the Walker class, but for that little bit, should I create a subclass?

In the end I found the not so ideal solution, which you can find in this gist: https://gist.github.com/iljester/77eeb23a5dd19e1fd98ba535758fb984

Hence my proposal. I repeat: this cannot be done for all translatable strings, but I believe that whatever ends up in the output of the frontend of the site/blog must be easily customizable in some way.

**Comment by KTS915** [comment]
It's quite possible to add html to the translated string, which then creates possibilities to add an icon (whether via CSS or in the HTML itself). But I'm not necessarily against the `add_action` hook; I just don't understand why the function with the filter should be in core.

As @ginsterbusch suggested, can't you put that function in your own theme or plugin? Then the relevant PR should just add the action in core.

**Comment by iljester** [comment]
I know perfectly well that it could also be done via css. But you still need to translate the string. If this were the case then we would also have had to entrust many other cases of output with filters and actions to this system. Why introduce actions and filters if you could then solve everything with gettext?

However, I made the proposal. It is up to core development to decide how and whether to do this. I accept any decision. Even just `add_action` would be fine. Then the plugin or theme would do the rest.

**Comment by KTS915** [comment]
I took a further look at this with a view to writing a PR, but then I realized that it's going to be a breaking change. This is because it involves moving the `$moderation_note` from within `<footer>` to within `<div class="comment-content">`. That has the potential to mess up the presentation on any theme. So, as it stands, this can't be included until version 3.0 at the earliest.

In fact, including just the `add_action` hook will also be a breaking change even if `$moderation_note` isn't moved to a different code block, because the moderation note then won't show without an additional function to display it.

If someone wants to suggest how this might be included in CP without making a breaking change, then we could look at including it in version 2.x. Otherwise, this is going to have to wait for a while.

**Comment by pattonwebz** [comment]
I still feel like subclassing the walker, overriding just one or 2 methods, in a theme is the most ideal solution here. No core changes needed and then the theme has full control of the positioning, markup and generation of it.

**Comment by ginsterbusch** [comment]
One could actually just add an action hook to the appropriate section (ie. `<div class="comment-content">`, and then either add another one, on the original section in the footer, or just a simple if-statement, that checks if this first action hook has executed a callback successfully (or eg. if a constant is set), and if not, display the original code / html block. That one could even be enhanced and turned into a pluggable function.


Just my .02 euro cents :)

cu, w0lf.

**Comment by iljester** [comment]
Good idea @ginsterbusch.

 In any case thanks for your answers. If you think it's okay to subclass the Walker (which I think is a less immediate and clean system, especially in the context of developing a theme), you can easily close the proposal here, and perhaps if the development team still considers it a valid proposal, postpone it - as Tim says - to a major upgrade.

I would like to point out, however, that the moderation note should not be between the `<footer></footer>` tags, at least in my opinion. But it's just an opinion.

**Comment by mattyrob** [comment]
Customisation of string translation is possible everywhere via the `gettext` hook:
This code works locally for me, see before and after images below.
```
add_filter( 'gettext', 'customize_comment_moderation_notice', 10, 3 );

function customize_comment_moderation_notice( $translated_text, $original_text, $domain ) {
	$moderation_note = 'Your comment is awaiting moderation.';
	$preview_moderation_note = 'Your comment is awaiting moderation. This is a preview; your comment will be visible after it has been approved.';
	
	if ( $original_text === $moderation_note ) {
		$translated_text = 'This comment will be visible after it has been approved by a moderator.';
	}
	
	if ( $original_text === $preview_moderation_note ) {
		$translated_text = 'This is a preview; your comment will be visible after it has been approved by a moderator.';
	}

	return $translated_text;
}
```

Before:
![Screenshot 2024-07-10 at 15 18 26](https://github.com/ClassicPress/ClassicPress/assets/1280733/bfd85999-f314-4412-b047-f5d6ba0b26e4)


After:
![Screenshot 2024-07-10 at 15 18 11](https://github.com/ClassicPress/ClassicPress/assets/1280733/cabc7927-4e1a-46d8-ab89-a5aeefa168a1)

**Comment by pattonwebz** [comment]
> Customisation of string translation is possible everywhere via the `gettext` hook

The requested hooks is less for modifying the text and more about being able to modify the surrounding markup which currently isn't achievable via hooks. Instead requires a custom walker since no hooks are present for it, which I personally feel is a suitable solve - but has more overhead than a simple hooks set.

**Comment by mattyrob** [comment]
> The requested hooks is less for modifying the text and more about being able to modify the surrounding markup which currently isn't achievable via hooks. Instead requires a custom walker since no hooks are present for it, which I personally feel is a suitable solve - but has more overhead than a simple hooks set.

Ah, I see that now having read the comments - it's not immediately apparent to me from the original request nor example code.

**Comment by KTS915** [comment]
> The requested hooks is less for modifying the text and more about being able to modify the surrounding markup which currently isn't achievable via hooks.

I've looked further at this since I commented above. Actually, it _is_ achievable via the `gettext` filter. You can add whatever markup you like using that filter. Just set `$domain` to `'default'` so that the translation is attached only to this hook (and maybe be careful with the priority if you think that something else might also be attempting to translate the same string).

**Comment by mattyrob** [comment]
On the premise that a new hook / filter is needed to allow full control over the output of the `html5_comment()` function in the comment walker class, this could be achieved with core changes as follows I think. The need for such a new filter is of course up for debate still, but....

In the `start_el()` function in the same file the html5_comment()` function is called as follows:
```
ob_start();
$this->html5_comment( $comment, $depth, $args );
$output .= ob_get_clean();
```

The hook can be implemented by filtering the completed reutrn from the function with the following:
```
ob_start();
$this->html5_comment( $comment, $depth, $args );
$buffer = ob_get_clean();

$output .= apply_filters( 'html5_comment', $buffer );
```

**Comment by KTS915** [comment]
Fixed by PR #1518.
